packages/jsii/lib/transforms/deprecation-warnings-transformer.ts

```diff
@@ -0,0 +1,618 @@
+import {
+  Assembly,
+  ClassType,
+  InterfaceType,
+  isClassOrInterfaceType,
+  Stability,
+} from '@jsii/spec';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as ts from 'typescript';
+
+import * as bindings from '../node-bindings';
+import {
+  fullyQualifiedName,
+  isDeprecated,
+  isInternal,
+  isPublic,
+} from './utils';
+
+const WARNING_FUNCTION_NAME = 'printJsiiDeprecationWarnings';
+const FILE_NAME = '.warnings.jsii.js';
+
+type MethodLikeDeclaration =
+  | ts.MethodDeclaration
+  | ts.ConstructorDeclaration
+  | ts.GetAccessorDeclaration
+  | ts.SetAccessorDeclaration;
+
+interface Warning {
+  readonly elementName: string;
+  readonly message: string;
+  readonly path?: string;
+}
+
+/*
+
+General idea:
+
+When the JSII assembly is created, DeprecatedWarningInjector.process should be
+invoked so that it can compute and store general information about the module,
+such as deprecated types and class and interface declarations. It also creates a
+new Javascript file with a function that prints warnings.
+
+At emit time, DeprecatedWarningsTransformer.transform is invoked for each source
+file and then recursively for each node in the AST. We are interested in two
+types of nodes: class declarations and method-like declarations (proper methods,
+constructors and accessors).
+
+For classes, we navigate up in the type hierarchy looking for deprecated types.
```

Comment:
Wait. I'm not sure we need to do that. 

I don't think it's true that if a class' base class is deprecated, it automatically means the subclass is deprecated.